### PR TITLE
Enable pip installation for Marius 

### LINKED ISSUE #1: Make Marius pip installable
**Is your feature request related to a problem? Please describe.**
Users shouldn't have to build Marius with cmake to use it. We should provide pip install capabilities to simplify the usage/  

**Describe the solution you'd like**
Building and install Marius from source:
```
git clone https://github.com/marius-team/marius.git
cd marius
python3 -m pip install .
```

Installing Marius from PyPi:
```
python3 -m pip install pymarius 
// or 
python3 -m pip install marius 
```
**Describe alternatives you've considered**
The alternative is to make people build it themselves from source with the instructions in the docs. Not great...

**Additional context**
Should we separate out installing the Python API vs. the config based executable?

### PR DESCRIPTION
**Pull request by JasonMoho**
Fixes #1 

This change enables users to build and install Marius with pip, instead of using CMake.

By running `pip install .` in the root directory of the repository the following will occur:
- `marius_train` executable will be installed on the user's system path
- `marius_eval` executable will be installed on the user's system path
- `marius_preprocess` entry point to tools/preprocessing.py is added to the users system path.
- `marius_postprocess` entry point to tools/postprocess.py is added to the users system path.
- The python bindings become available for import in Python. E.g `import marius as m`

This change includes an additional workflow which tests the pip install on linux and MacOS.

### CODE REVIEW

File: include/util.h [hunk 1]
```diff
@@ -18,33 +18,35 @@ class Timer {
     bool gpu_;
     std::chrono::time_point<std::chrono::high_resolution_clock> start_time_;
     std::chrono::time_point<std::chrono::high_resolution_clock> stop_time_;
-    CudaEvent start_event_;
-    CudaEvent end_event_;
+    CudaEvent *start_event_;
```
Comment (shivaram) [inline include/util.h:21]: these changes dont seem related to pip?

File: python_bindings/batch_wrap.cpp [hunk 2]
```diff
@@ -53,8 +53,6 @@ void init_batch(py::module &m) {
         .def_readwrite("dst_all_neg_embeddings_", &Batch::dst_all_neg_embeddings_)
         .def_readwrite("load_timestamp_", &Batch::load_timestamp_)
         .def_readwrite("compute_timestamp_", &Batch::compute_timestamp_)
-	    .def_readwrite("device_transfer_", &Batch::device_transfer_)
```
Comment (shivaram) [inline python_bindings/batch_wrap.cpp:56]: also doesn't seem related to pip

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        print(cmake_args)
+
+        subprocess.check_call(['cmake', ext.sourcedir] + cmake_args,
+                              cwd=self.build_temp, env=env)
+        subprocess.check_call(['cmake', '--build', '.'] + build_args,
+                              cwd=self.build_temp)
+        print()  # Add an empty line for cleaner output
+
+
+only_python = os.environ.get("MARIUS_ONLY_PYTHON", None)
```
Comment (shivaram) [inline setup.py:76]: just curious -- what does MARIUS_ONLY_PYTHON do?

**Review by shivaram — COMMENTED** [review]
This is great! Some minor comments

File: include/util.h (hunk 1 above)
Comment (JasonMoho) [reply include/util.h:21]: This is to fix a bug with building the bindings of the GPU version of Marius. Came across it while testing the pip install and bundled it in with this pull request.

File: python_bindings/batch_wrap.cpp (hunk 2 above)
Comment (JasonMoho) [reply python_bindings/batch_wrap.cpp:56]: Same as above

File: setup.py (hunk 3 above)
Comment (JasonMoho) [reply setup.py:76]: This will only install marius.tools and won't build Marius. This is useful to save build/install time on tests which only test the preprocessing/postprocessing code.

File: include/util.h (hunk 1 above)
Comment (shivaram) [reply include/util.h:21]: Its generally better to have a separate PR for bug fixes -- makes it easier to track down later if you are looking at history to figure out when this was fixed and how etc.

Comment (JasonMoho) [reply include/util.h:21]: Make sense, I'll avoid doing this in future PRs

**Review by AnzeXie — COMMENTED** [review]
Everything looks good!
